File: backend/app/repositories/conversations.py
"""Tenant-scoped conversations and messages."""

from __future__ import annotations

from datetime import datetime, timezone
from uuid import UUID

from sqlalchemy import delete, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.entities import Conversation, ConversationDocument, Message
from app.models.enums import MessageRole, MessageStatus


class ConversationRepository:
    """Conversation persistence with ownership filters on every operation."""

    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    async def create(self, user_id: UUID, title: str, document_ids: list[UUID]) -> Conversation:
        """Create a conversation and bind selected tenant documents."""
        conversation = Conversation(user_id=user_id, title=title[:255])
        self.session.add(conversation)
        await self.session.flush()
        self.session.add_all(
            [
                ConversationDocument(
                    user_id=user_id,
                    conversation_id=conversation.id,
                    document_id=document_id,
                )
                for document_id in dict.fromkeys(document_ids)
            ]
        )
        await self.session.flush()
        return conversation

    async def get(self, user_id: UUID, conversation_id: UUID) -> Conversation | None:
        """Find one tenant conversation."""
        result = await self.session.execute(
            select(Conversation).where(
                Conversation.id == conversation_id,
                Conversation.user_id == user_id,
                Conversation.deleted_at.is_(None),
            )
        )
        return result.scalar_one_or_none()

    async def replace_documents(
        self, user_id: UUID, conversation_id: UUID, document_ids: list[UUID]
    ) -> None:
        """Update documents available to a conversation."""
        await self.session.execute(
            delete(ConversationDocument).where(
                ConversationDocument.user_id == user_id,
                ConversationDocument.conversation_id == conversation_id,
            )
        )
        self.session.add_all(
            [
                ConversationDocument(
                    user_id=user_id,
                    conversation_id=conversation_id,
                    document_id=document_id,
                )
                for document_id in dict.fromkeys(document_ids)
            ]
        )
        await self.session.flush()

    async def list(
        self, user_id: UUID, *, limit: int, offset: int
    ) -> tuple[list[Conversation], int]:
        """List tenant conversations."""
        filters = (Conversation.user_id == user_id, Conversation.deleted_at.is_(None))
        items = await self.session.execute(
            select(Conversation)
            .where(*filters)
            .order_by(Conversation.updated_at.desc())
            .limit(limit)
            .offset(offset)
        )
        total = await self.session.execute(select(func.count(Conversation.id)).where(*filters))
        return list(items.scalars().all()), int(total.scalar_one())

    async def add_message(
        self,
        *,
        user_id: UUID,
        conversation_id: UUID,
        role: MessageRole,
        content: str,
        status: MessageStatus = MessageStatus.COMPLETED,
    ) -> Message:
        """Persist one message."""
        message = Message(
            user_id=user_id,
            conversation_id=conversation_id,
            role=role,
            content=content,
            status=status,
        )
        self.session.add(message)
        await self.session.flush()
        return message

    async def list_messages(
        self, user_id: UUID, conversation_id: UUID, *, limit: int, offset: int
    ) -> tuple[list[Message], int]:
        """List non-deleted messages only after tenant ownership is established."""
        conversation = await self.get(user_id, conversation_id)
        if conversation is None:
            return [], 0
        filters = (
            Message.user_id == user_id,
            Message.conversation_id == conversation_id,
            Message.deleted_at.is_(None),
        )
        items = await self.session.execute(
            select(Message).where(*filters).order_by(Message.created_at).limit(limit).offset(offset)
        )
        total = await self.session.execute(select(func.count(Message.id)).where(*filters))
        return list(items.scalars().all()), int(total.scalar_one())

    async def get_message(
        self, user_id: UUID, conversation_id: UUID, message_id: UUID
    ) -> Message | None:
        """Find one tenant-owned message."""
        result = await self.session.execute(
            select(Message).where(
                Message.id == message_id,
                Message.user_id == user_id,
                Message.conversation_id == conversation_id,
                Message.deleted_at.is_(None),
            )
        )
        return result.scalar_one_or_none()

    async def complete_message(
        self,
        message: Message,
        *,
        content: str,
        citations: list[dict[str, object]],
        model_name: str,
        latency_ms: int,
    ) -> None:
        """Finalize an assistant message."""
        message.content = content
        message.citations = citations
        message.model_name = model_name
        message.latency_ms = latency_ms
        message.status = MessageStatus.COMPLETED
        await self.session.flush()

    async def fail_message(self, message: Message, error: str) -> None:
        """Mark a generation as failed without exposing provider internals."""
        message.status = MessageStatus.FAILED
        message.error_message = error[:4000]
        await self.session.flush()

    async def soft_delete(self, conversation: Conversation) -> None:
        """Soft delete a conversation and its visible messages."""
        now = datetime.now(timezone.utc)
        conversation.deleted_at = now
        await self.session.execute(
            Message.__table__.update()
            .where(
                Message.user_id == conversation.user_id,
                Message.conversation_id == conversation.id,
            )
            .values(deleted_at=now)
        )
        await self.session.flush()

File: backend/src/repositories/conversations.js
import crypto from "node:crypto";

export async function createConversation(client, userId, title, documentIds) {
  const conversationId = crypto.randomUUID();
  const result = await client.query(
    `
      INSERT INTO conversations (id, user_id, title)
      VALUES ($1, $2, $3)
      RETURNING *
    `,
    [conversationId, userId, title.slice(0, 255)],
  );
  await replaceConversationDocuments(client, userId, conversationId, documentIds);
  return result.rows[0];
}

export async function getConversation(client, userId, conversationId) {
  const result = await client.query(
    `
      SELECT *
      FROM conversations
      WHERE id = $1 AND user_id = $2 AND deleted_at IS NULL
    `,
    [conversationId, userId],
  );
  return result.rows[0] ?? null;
}

export async function replaceConversationDocuments(client, userId, conversationId, documentIds) {
  await client.query(
    "DELETE FROM conversation_documents WHERE user_id = $1 AND conversation_id = $2",
    [userId, conversationId],
  );
  for (const documentId of [...new Set(documentIds)]) {
    await client.query(
      `
        INSERT INTO conversation_documents (id, user_id, conversation_id, document_id)
        VALUES ($1, $2, $3, $4)
      `,
      [crypto.randomUUID(), userId, conversationId, documentId],
    );
  }
}

export async function listConversations(client, userId, { limit, offset }) {
  const [items, total] = await Promise.all([
    client.query(
      `
        SELECT id, title, created_at, updated_at
        FROM conversations
        WHERE user_id = $1 AND deleted_at IS NULL
        ORDER BY updated_at DESC
        LIMIT $2 OFFSET $3
      `,
      [userId, limit, offset],
    ),
    client.query(
      `
        SELECT count(*)::int AS total
        FROM conversations
        WHERE user_id = $1 AND deleted_at IS NULL
      `,
      [userId],
    ),
  ]);
  return { items: items.rows, total: total.rows[0].total };
}

export async function addMessage(
  client,
  { userId, conversationId, role, content, status = "COMPLETED" },
) {
  const result = await client.query(
    `
      INSERT INTO messages (id, user_id, conversation_id, role, status, content)
      VALUES ($1, $2, $3, $4, $5, $6)
      RETURNING *
    `,
    [crypto.randomUUID(), userId, conversationId, role, status, content],
  );
  await client.query(
    "UPDATE conversations SET updated_at = now() WHERE id = $1 AND user_id = $2",
    [conversationId, userId],
  );
  return result.rows[0];
}

export async function getMessage(client, userId, conversationId, messageId) {
  const result = await client.query(
    `
      SELECT *
      FROM messages
      WHERE id = $1 AND user_id = $2 AND conversation_id = $3 AND deleted_at IS NULL
    `,
    [messageId, userId, conversationId],
  );
  return result.rows[0] ?? null;
}

export async function listMessages(client, userId, conversationId, { limit, offset }) {
  const [items, total] = await Promise.all([
    client.query(
      `
        SELECT id, conversation_id, role, status, content, citations, model_name, created_at
        FROM messages
        WHERE user_id = $1 AND conversation_id = $2 AND deleted_at IS NULL
        ORDER BY created_at
        LIMIT $3 OFFSET $4
      `,
      [userId, conversationId, limit, offset],
    ),
    client.query(
      `
        SELECT count(*)::int AS total
        FROM messages
        WHERE user_id = $1 AND conversation_id = $2 AND deleted_at IS NULL
      `,
      [userId, conversationId],
    ),
  ]);
  return { items: items.rows, total: total.rows[0].total };
}

export async function completeMessage(
  client,
  messageId,
  { content, citations, modelName, latencyMs },
) {
  await client.query(
    `
      UPDATE messages
      SET content = $2,
          citations = $3::jsonb,
          model_name = $4,
          latency_ms = $5,
          status = 'COMPLETED',
          updated_at = now()
      WHERE id = $1
    `,
    [messageId, content, JSON.stringify(citations), modelName, latencyMs],
  );
}

export async function failMessage(client, messageId, error) {
  await client.query(
    `
      UPDATE messages
      SET status = 'FAILED', error_message = $2, updated_at = now()
      WHERE id = $1
    `,
    [messageId, String(error).slice(0, 4000)],
  );
}

export async function softDeleteConversation(client, userId, conversationId) {
  await client.query(
    `
      UPDATE conversations
      SET deleted_at = now(), updated_at = now()
      WHERE id = $1 AND user_id = $2
    `,
    [conversationId, userId],
  );
  await client.query(
    `
      UPDATE messages
      SET deleted_at = now(), updated_at = now()
      WHERE conversation_id = $1 AND user_id = $2
    `,
    [conversationId, userId],
  );
}

export function publicConversation(conversation) {
  return {
    id: conversation.id,
    title: conversation.title,
    created_at: new Date(conversation.created_at).toISOString(),
    updated_at: new Date(conversation.updated_at).toISOString(),
  };
}

export function publicMessage(message) {
  return {
    id: message.id,
    conversation_id: message.conversation_id,
    role: message.role,
    status: message.status,
    content: message.content,
    citations: message.citations ?? [],
    model_name: message.model_name,
    created_at: new Date(message.created_at).toISOString(),
  };
}
